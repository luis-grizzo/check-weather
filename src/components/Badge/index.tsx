interface BadgeProps {
  children: string
}

export function Badge({ children }: BadgeProps) {
  return (
    <span className="flex items-center gap-2 px-2 py-1 text-xs bg-neutral-100/60 border-1 border-neutral-600/10 text-neutral-950 rounded-lg backdrop-blur">
      {children}
    </span>
  )
}
