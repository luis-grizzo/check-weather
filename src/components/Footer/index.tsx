'use client'

export function Footer() {
  const currentYear = new Date().getFullYear()

  return (
    <footer className="bg-neutral-100/60 border-t-1 border-neutral-600/10">
      <p className="container mx-auto px-8 py-6 text-xs text-center">
        Developed by{' '}
        <a
          href="https://luisgrizzo.dev"
          target="_blank"
          rel="noreferrer"
          className="text-blue-600 hover:text-blue-400 transition-colors"
        >
          Lu√≠s Grizzo
        </a>
        {', '}
        <a
          href="https://github.com/luis-grizzo/check-weather?tab=MIT-1-ov-file"
          target="_blank"
          className="text-xs text-blue-600 hover:text-blue-400 transition-colors"
          rel="noreferrer"
        >
          MIT License
        </a>
        {', '}
        {currentYear}.
      </p>
    </footer>
  )
}
