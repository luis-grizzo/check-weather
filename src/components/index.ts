export * from './Badge'
export * from './Button'
export * from './DisplayTime'
export * from './Footer'
export * from './Header'
export * from './InfoPanel'
export * from './Navbar'
export * from './RefreshButton'
