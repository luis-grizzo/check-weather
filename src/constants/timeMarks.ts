export const hourInMilliseconds = 3_600_000

export const hourInSeconds = 3_600
