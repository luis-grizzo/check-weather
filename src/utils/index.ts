export * from './numberFormaters'
export * from './randomIcon'
export * from './stringFormaters'
