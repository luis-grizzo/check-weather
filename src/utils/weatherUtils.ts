import {
  formatCountryName,
  formatDate,
  formatTime
} from '@/lib/stringFormaters'

import {
  calculatePeriod,
  calculateVisibility,
  calculateWindSpeed
} from '@/lib/calculators'

import { roundDown, roundUp, roundValue } from '@/lib/numberFormaters'

import type { FetchWeatherResponse } from '@/services/fetchWeather'
import type { GenericWeatherInfo, WeatherTypes } from '@/types/weatherTypes'

export interface FetchWeatherFactoredResponse {
  type: WeatherTypes
  description: string
  temperature: string
  forecast_temperatures: GenericWeatherInfo[]
  time: string
  unixTimestamp: number
  forecast_times: GenericWeatherInfo[]
  period: 'day' | 'night'
  visibility: string
  wind_speed: string
  humidity: string
  location: {
    city: string
    country?: string
  }
}

export function formatWeatherResponse(
  response: FetchWeatherResponse
): FetchWeatherFactoredResponse {
  return {
    type: response.weather[0].main.toLowerCase() as FetchWeatherFactoredResponse['type'],
    description: response.weather[0].description,
    temperature: `${roundValue(response.main.temp)}째`,
    forecast_temperatures: [
      {
        description: 'Feels like',
        value: `${roundValue(response.main.feels_like)}째`
      },
      {
        description: 'Min',
        value: `${roundDown(response.main.temp_min)}째`
      },
      {
        description: 'Max',
        value: `${roundUp(response.main.temp_max)}째`
      }
    ],
    time: `${formatTime(response.dt)} - ${formatDate(response.dt)}`,
    unixTimestamp: response.dt,
    forecast_times: [
      {
        description: 'Sunrise',
        value: formatTime(response.sys.sunrise)
      },
      {
        description: 'Sunset',
        value: formatTime(response.sys.sunset)
      }
    ],
    period: calculatePeriod(
      response.dt,
      response.sys.sunrise,
      response.sys.sunset
    ),
    visibility: calculateVisibility(response.sys.country, response.visibility),
    wind_speed: calculateWindSpeed(response.sys.country, response.wind.speed),
    humidity: `${response.main.humidity}%`,
    location: {
      city: response.name,
      country: formatCountryName(response.sys.country)
    }
  }
}
