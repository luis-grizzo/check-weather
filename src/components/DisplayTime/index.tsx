'use client'

import { Badge } from '@/components/Badge'

import { formatTime } from '@/lib/stringFormaters'

interface DisplayTimeProps {
  description: string
  timestamp: number
}

export function DisplayTime({ description, timestamp }: DisplayTimeProps) {
  return <Badge>{`${description} ${formatTime(timestamp)}`}</Badge>
}
