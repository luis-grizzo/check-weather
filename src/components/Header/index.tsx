'use client'

import { useEffect, useState } from 'react'

import { hourInMilliseconds } from '@/constants/timeMarks'

import {
  formatCountryName,
  formatDate,
  formatTime,
  formatTimer
} from '@/lib/stringFormaters'
import { positiveOrZero } from '@/lib/numberFormaters'

import type { FetchWeatherFactoredResponse } from '@/utils/weatherUtils'
import { useRouter } from 'next/navigation'

interface HeaderProps {
  location: FetchWeatherFactoredResponse['location']
  requestTimestamp: number
}

export function Header({ location, requestTimestamp }: HeaderProps) {
  const router = useRouter()

  const [countdownTimer, setCountdownTimer] = useState('--:--')
  const [shouldRefresh, setShouldRefresh] = useState(false)

  useEffect(() => {
    const timer = setInterval(() => {
      const currentTime = Date.now()
      const revalidateTime = requestTimestamp * 1_000 + hourInMilliseconds

      const remainTime = revalidateTime - currentTime
      const validatedRemainTime = positiveOrZero(remainTime)

      const formattedTimer = formatTimer(validatedRemainTime)

      setCountdownTimer(formattedTimer)

      if (!validatedRemainTime) {
        setShouldRefresh(true)

        clearInterval(timer)
      }
    }, 1_000)

    return () => clearInterval(timer)
  }, [requestTimestamp])

  useEffect(() => {
    if (shouldRefresh) {
      router.refresh()

      setShouldRefresh(false)
    }
  }, [shouldRefresh])

  return (
    <header className="flex items-center justify-between container mx-auto px-8 py-6">
      <div className="flex flex-col gap-1">
        <span role="timer" className="text-xs">
          {`Next forecast in ${countdownTimer}`}
        </span>

        {location.country && (
          <span className="text-lg">
            {`${location.city}, ${formatCountryName(location.country)}`}
          </span>
        )}

        <span className="text-xs">{`${formatTime(requestTimestamp)} - ${formatDate(requestTimestamp)}`}</span>
      </div>
    </header>
  )
}
