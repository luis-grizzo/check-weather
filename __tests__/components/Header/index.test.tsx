import { act } from 'react'
import { render, screen, waitFor } from '@testing-library/react'

import { hourInMilliseconds } from '@/constants/timeMarks'

import { Header } from '@/components'

import { FetchWeatherFactoredResponse } from '@/utils/weatherUtils'

const unixTimestamp = new Date('June 17, 2024 16:00').getTime() / 1_000
const secondInMilliseconds = 1_000

jest.useFakeTimers()

const refreshMock = jest.fn()

jest.mock('next/navigation', () => ({
  useRouter() {
    return {
      refresh: refreshMock
    }
  }
}))

const location: FetchWeatherFactoredResponse['location'] = {
  city: 'JaÃº',
  country: 'BR'
}

describe('Header', () => {
  beforeEach(() => {
    jest.setSystemTime(
      new Date(unixTimestamp * 1_000 + hourInMilliseconds - 5_000)
    )

    render(<Header location={location} requestTimestamp={unixTimestamp} />)
  })

  afterEach(() => {
    jest.clearAllTimers()
  })

  it('should match snapshot', () => {
    const { container } = render(
      <Header location={location} requestTimestamp={unixTimestamp} />
    )

    expect(container).toMatchSnapshot()
  })

  it('should update the timer', async () => {
    let eTimer = screen.getByRole('timer')

    expect(eTimer.textContent).toEqual('Next forecast in --:--')

    act(() => {
      jest.advanceTimersByTime(secondInMilliseconds)
    })

    await waitFor(() => {
      eTimer = screen.getByRole('timer')

      expect(eTimer.textContent).toEqual('Next forecast in 00:04')
    })
  })

  it('should call refresh when timer reaches zero', async () => {
    act(() => {
      jest.advanceTimersByTime(5_000)
    })

    await waitFor(() => {
      expect(refreshMock).toHaveBeenCalled()
    })
  })

  it('should timer not be minor than zero', () => {
    act(() => {
      jest.advanceTimersByTime(10_000)
    })

    const eTimer = screen.getByRole('timer')

    expect(eTimer.textContent).toEqual('Next forecast in 00:00')
  })
})
