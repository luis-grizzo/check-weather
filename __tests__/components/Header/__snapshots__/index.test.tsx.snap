// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should match snapshot 1`] = `
<div>
  <header
    class="flex items-center justify-between container mx-auto px-8 py-6"
  >
    <div
      class="flex flex-col gap-1"
    >
      <span
        class="text-xs"
        role="timer"
      >
        Next forecast in --:--
      </span>
      <span
        class="text-lg"
      >
        Ja√∫, Brazil
      </span>
      <span
        class="text-xs"
      >
        4:00 PM - Monday, June 17, 2024
      </span>
    </div>
  </header>
</div>
`;
